### PR TITLE
[core] add `Exclusive` to sync

### PR DESCRIPTION
**Pull request by guswynn**
(discussed here: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Adding.20.60SyncWrapper.60.20to.20std)

`Exclusive` is a wrapper that exclusively allows mutable access to the inner value if you have exclusive access to the wrapper. It acts like a compile time mutex, and hold an unconditional `Sync` implementation.

## Justification for inclusion into std
- This wrapper unblocks actual problems:
  - The example that I hit was a vector of `futures::future::BoxFuture`'s causing a central struct in a script to be non-`Sync`. To work around it, you either write really difficult code, or wrap the futures in a needless mutex.
- Easy to maintain: this struct is as simple as a wrapper can get, and its `Sync` implementation has very clear reasoning
- Fills a gap: `&/&mut` are to `RwLock` as `Exclusive` is to `Mutex`

## Public Api
```rust
// core::sync
#[derive(Default)]
struct Exclusive<T: ?Sized> { ... }

impl<T: ?Sized> Sync for Exclusive {}

impl<T> Exclusive<T> {
    pub const fn new(t: T) -> Self;
    pub const fn into_inner(self) -> T;
}

impl<T: ?Sized> Exclusive<T> {
    pub const fn get_mut(&mut self) -> &mut T;
    pub const fn get_pin_mut(Pin<&mut self>) -> Pin<&mut T>;
    pub const fn from_mut(&mut T) -> &mut Exclusive<T>;
    pub const fn from_pin_mut(Pin<&mut T>) -> Pin<&mut Exclusive<T>>;
}

impl<T: Future> Future for Exclusive { ... }

impl<T> From<T> for Exclusive<T> { ... }
impl<T: ?Sized> Debug for Exclusive { ... }
```

## Naming
This is a big bikeshed, but I felt that `Exclusive` captured its general purpose quite well.

## Stability and location
As this is so simple, it can be in `core`. I feel that it can be stabilized quite soon after it is merged, if the libs teams feels its reasonable to add. Also, I don't really know how unstable feature work in std/core's codebases, so I might need help fixing them

## Tips for review
The docs probably are the thing that needs to be reviewed! I tried my best, but I'm sure people have more experience than me writing docs for `Core`

### Implementation:
The API is mostly pulled from https://docs.rs/sync_wrapper/latest/sync_wrapper/struct.SyncWrapper.html (which is apache 2.0 licenesed), and the implementation is trivial:
- its an unsafe justification for pinning
- its an unsafe justification for the `Sync` impl (mostly reasoned about by @danielhenrymantilla here: https://github.com/Actyx/sync_wrapper/pull/2)
- and forwarding impls, starting with derivable ones and `Future`

### CODE REVIEW

File: library/core/src/sync/exclusive.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+/// In some sense, `Exclusive` can be thought of as a _compile-time_ version of
+/// a mutex, as the borrow-checker guarantees that only one `&mut` can exist
+/// for any value. This is a parallel with the fact that
+/// `&` and `&mut` references together can be thought of as a _compile-time_
+/// version of a read-write lock.
+///
+///
+/// [`Sync`]: core::marker::Sync
+#[unstable(feature = "exclusive_struct", issue = "none")]
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
```
Comment (cuviper) [inline library/core/src/sync/exclusive.rs:73]: Only `Default` is safe here, as the rest will derive access from `&Self` to `&T`. We could have a manual `Debug` that doesn't print the inner value, but none of the others are feasible.

**Review by cuviper — CHANGES_REQUESTED** [review]

File: library/core/src/sync/exclusive.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    /// access to the underlying value, but _pinned_ `Exclusive`s only
+    /// produce _pinned_ access to the underlying value.
+    #[unstable(feature = "exclusive_struct", issue = "none")]
+    pub fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut T> {
+        // SAFETY: `Exclusive` can only produce `&mut T` if itself is unpinned
+        unsafe { Pin::new_unchecked(&mut self.get_unchecked_mut().inner) }
+    }
+}
+
+// See `Exclusive`'s docs for justification.
```
Comment (ChayimFriedman2) [inline library/core/src/sync/exclusive.rs:115]: How is that relevant here?

File: library/core/src/sync/exclusive.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+#[unstable(feature = "exclusive_struct", issue = "none")]
+impl<T> From<T> for Exclusive<T> {
+    fn from(t: T) -> Self {
+        Self { inner: t }
```
Comment (ChayimFriedman2) [inline library/core/src/sync/exclusive.rs:119]: Can use `new()` here: this is what for example [`UnsafeCell` does](https://doc.rust-lang.org/stable/src/core/cell.rs.html#1983-1988).

File: library/core/src/sync/exclusive.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+/// a mutex, as the borrow-checker guarantees that only one `&mut` can exist
+/// for any value. This is a parallel with the fact that
+/// `&` and `&mut` references together can be thought of as a _compile-time_
+/// version of a read-write lock.
+///
+///
+/// [`Sync`]: core::marker::Sync
+#[unstable(feature = "exclusive_struct", issue = "none")]
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+#[repr(transparent)]
```
Comment (ChayimFriedman2) [inline library/core/src/sync/exclusive.rs:74]: Why is this `#[repr(transparent)]`?

File: library/core/src/sync/exclusive.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+// See `Exclusive`'s docs for justification.
+#[unstable(feature = "exclusive_struct", issue = "none")]
+unsafe impl<T> Sync for Exclusive<T> {}
+
+impl<T> Exclusive<T> {
+    /// Wrap a value in an `Exclusive`
+    #[unstable(feature = "exclusive_struct", issue = "none")]
+    pub fn new(t: T) -> Self {
```
Comment (SabrinaJewson) [inline library/core/src/sync/exclusive.rs:86]:
```suggestion
    pub const fn new(t: T) -> Self {
```

Comment (SabrinaJewson) [inline library/core/src/sync/exclusive.rs:86]:
```suggestion
    #[must_use]
    pub fn new(t: T) -> Self {
```

Since it has no side-effects

File: library/core/src/sync/exclusive.rs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        Self { inner: t }
+    }
+
+    /// Unwrap the value contained in the `Exclusive`
+    #[unstable(feature = "exclusive_struct", issue = "none")]
+    pub fn into_inner(self) -> T {
```
Comment (SabrinaJewson) [inline library/core/src/sync/exclusive.rs:92]: `#[must_use]` also makes sense here, as well as `get_mut` and `get_pin_mut`. And make thiss `const` too.

File: library/core/src/sync/exclusive.rs (hunk 4 above)
Comment (danielhenrymantilla) [reply library/core/src/sync/exclusive.rs:74]: Given the design, there is no reason not to be, ever, so if we can provide this guarantee it could be nice for other code. But then I think it should be accompanied by the utility constructor:

```rust
// and-or a `From<&'lt mut T> for &'lt mut Exclusive<T>`
impl<T : ?Sized> Exclusive<T> {
    pub fn from_mut(r: &'_ mut T) -> &'_ mut Exclusive<T> {
        // SAFETY: repr is ≥ C, so refs have the same layout; and `Exclusive` properties are `&mut`-agnostic
        unsafe { &mut *(r as *mut T as *mut Exclusive<T>) }
    }
}
```

File: library/core/src/sync/exclusive.rs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+/// for any value. This is a parallel with the fact that
+/// `&` and `&mut` references together can be thought of as a _compile-time_
+/// version of a read-write lock.
+///
+///
+/// [`Sync`]: core::marker::Sync
+#[unstable(feature = "exclusive_struct", issue = "none")]
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+#[repr(transparent)]
+pub struct Exclusive<T> {
```
Comment (danielhenrymantilla) [inline library/core/src/sync/exclusive.rs:75]:
```suggestion
pub struct Exclusive<T : ?Sized> {
```

(see previous comment)

File: library/core/src/sync/exclusive.rs (hunk 4 above)
Comment (danielhenrymantilla) [reply library/core/src/sync/exclusive.rs:74]: And a `from_pin_mut` as well, I guess.

Both would be useful for the `[Pin<] &mut dyn Future [+ Unpin]` cases, where there would be no `Sync` _per se_ but we could end up needing it; we skip the layer of indirection that `Exclusive::new()` would have us go through 🙂

File: library/core/src/sync/exclusive.rs (hunk 1 above)
Comment (danielhenrymantilla) [reply library/core/src/sync/exclusive.rs:73]: 💯 – the `SyncWrapper` crate offered a dummy `Debug` for convenience, I believe we could do the same here?

**Review by danielhenrymantilla — COMMENTED** [review]
I like this proposal very much 🙂

**Comment by m-ou-se** [comment]
@guswynn Can you update the top comment to include a complete API overview? So just the types, methods, and trait impl blocks, without implementation or doc comments. (Just like we try to do on tracking issues.) That makes it a bit easier for reviewers to quickly get an idea of what API this PR is adding.

File: library/core/src/sync/exclusive.rs (hunk 4 above)
Comment (guswynn) [reply library/core/src/sync/exclusive.rs:74]: I think `from_mut` is nicer than adding another from impl, id rather keep the `From<T: ?Sized>` impl for the common case!

**Comment by guswynn** [comment]
@m-ou-se let me know if thats the format you wanted!

**Comment by guswynn** [comment]
@danielhenrymantilla  the future impl doesnt work with `?Sized` because of
```
error[E0599]: the method `get_pin_mut` exists for struct `Pin<&mut Exclusive<T>>`, but its trait bounds were not satisfied
   --> library/core/src/sync/exclusive.rs:163:14
    |
163 |         self.get_pin_mut().poll(cx)
    |              ^^^^^^^^^^^
    |
   ::: library/core/src/pin.rs:408:1
    |
408 | pub struct Pin<P> {
    | ----------------- method `get_pin_mut` not found for this
    |
    = note: the following trait bounds were not satisfied:
            `T: Sized`
```
which is unclear to me, should I add a `&mut Self: ?Sized` bound on the `get_pin_mut` method?

**Comment by danielhenrymantilla** [comment]
Yes, anything that operates through indirection can have a `?Sized` unbound on `T`; I suspect that only the `fn new()`, the `fn into_inner()`, and the `From` impl will be the ones requiring the (implicit) `T : Sized` bound

File: library/core/src/sync/exclusive.rs (hunk 4 above)
Comment (ChayimFriedman2) [reply library/core/src/sync/exclusive.rs:78]: As long as it is not documented, it is not considered a public guarantee.

File: library/core/src/sync/exclusive.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+    /// Build a _pinned mutable_ references to an `Exclusive<T>` from
+    /// a _pinned mutable_ reference to a `T`. This allows you to skip
+    /// building an `Exclusive` with [`Exclusive::new`].
+    #[unstable(feature = "exclusive_wrapper", issue = "none")]
+    #[must_use]
+    pub const fn from_pin_mut(r: Pin<&'_ mut T>) -> Pin<&'_ mut Exclusive<T>> {
+        // SAFETY: repr is ≥ C, so refs have the same layout; and `Exclusive` properties are `&mut`-agnostic
+        // SAFETY: `Exclusive` can only produce `&mut T` if itself is unpinned
+        unsafe { Pin::new_unchecked(&mut *(r.get_unchecked_mut() as *mut T as *mut Exclusive<T>)) }
```
Comment (ChayimFriedman2) [inline library/core/src/sync/exclusive.rs:149]: Maybe use `rom_mut()` here? This saves one safety comment :P

File: library/core/src/sync/exclusive.rs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    /// Build a _pinned mutable_ references to an `Exclusive<T>` from
+    /// a _pinned mutable_ reference to a `T`. This allows you to skip
+    /// building an `Exclusive` with [`Exclusive::new`].
+    #[unstable(feature = "exclusive_wrapper", issue = "none")]
+    #[must_use]
+    pub const fn from_pin_mut(r: Pin<&'_ mut T>) -> Pin<&'_ mut Exclusive<T>> {
+        // SAFETY: `Exclusive` can only produce `&mut T` if itself is unpinned
+        unsafe { Pin::new_unchecked(Self::from_mut(r.get_unchecked_mut())) }
```
Comment (ChayimFriedman2) [inline library/core/src/sync/exclusive.rs:151]: Looks like `Pin::map_unchecked_mut(r, Self::from_mut)`.

Comment (guswynn) [reply library/core/src/sync/exclusive.rs:151]: `map_unchecked_mut` is not const, but these are!

Comment (SabrinaJewson) [reply library/core/src/sync/exclusive.rs:151]: Sounds like it’d be good to add a comment explaining that then?
```suggestion
        // `Pin::map_unchecked_mut` is not const, so we do this conversion manually
        unsafe { Pin::new_unchecked(Self::from_mut(r.get_unchecked_mut())) }
```

File: library/core/src/sync/exclusive.rs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+use core::pin::Pin;
+use core::task::{Context, Poll};
+
+/// `Exclusive` provides only _mutable_ access, also referred to as _exclusive_
+/// accesss to the underlying value. It provides no _immutable_, or _shared_
+/// access to the underlying value.
+///
+/// While this may seem not very useful, it allows `Exclusive` to _unconditionally_
+/// implement [`Sync`]. Because _nothing_ can be done with an `&Exclusive`, it can
+/// trivially be send between threads, which satifies `Sync`'s safety requirements.
```
Comment (danielhenrymantilla) [inline library/core/src/sync/exclusive.rs:14]:
```suggestion
/// trivially be sent between threads, which satifies `Sync`'s safety requirements.
```

Comment (danielhenrymantilla) [inline library/core/src/sync/exclusive.rs:14]:
```suggestion
/// implement [`Sync`]. Indeed, the safety requirements of `Sync` state that for `Exclusive`
/// to be `Sync`, it must be sound to _share_ across threads, that is, it must be sound
/// for `&Exclusive` to cross thread boundaries. By design, a `&Exclusive` has no API
/// whatsoever, making it useless, thus harmless, thus memory safe.
```

File: library/core/src/sync/exclusive.rs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+///
+/// Certain constructs like [`Future`]s can only be used with _exclusive_ access,
+/// and are often `Send` but not `Sync`, so `Exclusive` can be used as hint to the
+/// rust compiler that something is `Sync` in practice.
+///
+/// ## Examples
+/// Using a non-`Sync` future prevent the wrapping struct from being `Sync`
```
Comment (danielhenrymantilla) [inline library/core/src/sync/exclusive.rs:21]:
```suggestion
/// Using a non-`Sync` future prevents the wrapping struct from being `Sync`:
```

File: library/core/src/sync/exclusive.rs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+///     future: async {
+///         let cell = Cell::new(1);
+///         let cell_ref = &cell;
+///         other().await;
+///         let value = cell_ref.get();
+///     }
+/// });
+/// ```
+///
+/// `Exclusive` ensure the struct is `Sync` without stripping the future of any
```
Comment (danielhenrymantilla) [inline library/core/src/sync/exclusive.rs:41]:
```suggestion
/// `Exclusive` ensures the struct is `Sync` without stripping the future of its
```

File: library/core/src/sync/exclusive.rs [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+#[unstable(feature = "exclusive_wrapper", issue = "none")]
+impl<T: Future + ?Sized> Future for Exclusive<T> {
+    type Output = T::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.get_pin_mut().poll(cx)
+    }
+}
```
Comment (danielhenrymantilla) [inline library/core/src/sync/exclusive.rs:171]: Could we have impls of `FnMut` and `FnOnce` as well?

Ditto for `Read`, `Write`, and basically any stdlib trait with `&mut self` methods exclusively

Comment (cuviper) [reply library/core/src/sync/exclusive.rs:171]: Note that `Read::is_read_vectored` and `Write::is_write_vectored` take `&self`, so those would both have to remain their default `false` rather than forwarding inward.

We _could_ generalize that principle to implement traits where all **required** methods are owned or mutable, leaving any `&self` methods to their trait default. That could include `Iterator`, for instance, implementing `next` but not `size_hint`. I'm not sure how far we should go with that, given that you can use `get_mut` or `into_inner` for direct access anyway.

**Comment by m-ou-se** [comment]
:robot: :mega: Team member @m-ou-se has proposed to merge this unstable api.

:bell: **This is now entering its \<unnamed\> comment period**, as per the review above. :bell:

No concerns currently listed.

(cc @rust-lang/libs-api)

File: library/core/src/sync/exclusive.rs (hunk 13 above)
Comment (cuviper) [reply library/core/src/sync/exclusive.rs:171]: I agree that `FnOnce` and `FnMut` are compelling for their simplicity and that they are unstable for users. It also related to [this forum thread](https://internals.rust-lang.org/t/dyn-fnonce-should-always-be-sync/16470/11), although I think they probably don't need the direct trait, per se.

File: library/core/src/sync/exclusive.rs [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+unsafe impl<T: ?Sized> Sync for Exclusive<T> {}
+
+#[unstable(feature = "exclusive_wrapper", issue = "none")]
+impl<T: ?Sized> fmt::Debug for Exclusive<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        f.debug_struct("Exclusive").finish_non_exhaustive()
+    }
+}
+
+impl<T: Sized> Exclusive<T> {
```
Comment (ChayimFriedman2) [inline library/core/src/sync/exclusive.rs:99]: Is there a reason for the explicit `: Sized` bound?

**Comment by joshtriplett** [comment]
@m-ou-se second

:grinning:

File: library/core/src/sync/exclusive.rs (hunk 14 above)
Comment (guswynn) [reply library/core/src/sync/exclusive.rs:99]: just to make it clear how its different than the other impl block

**Comment by guswynn** [comment]
@m-ou-se its been 10 days, is this merge-able?

**Comment by m-ou-se** [comment]
> @m-ou-se its been 10 days, is this merge-able?

Yes! Only five days late. We should probably replace me doing this manually by a bot that doesn't have adhd. :D

**Comment by m-ou-se** [comment]
:robot: :mega: The \<unnamed\> comment period, with a disposition to **merge**, as per the [review above](https://github.com/rust-lang/rust/pull/97629#issuecomment-1148917051), is now **complete**.

As the ~~automated~~ human representative of the governance process, I would like to thank the author for their work and everyone else who contributed.

This will be merged soon.

File: library/core/src/sync/exclusive.rs [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+/// ## Parallels with a mutex
+/// In some sense, `Exclusive` can be thought of as a _compile-time_ version of
+/// a mutex, as the borrow-checker guarantees that only one `&mut` can exist
+/// for any value. This is a parallel with the fact that
+/// `&` and `&mut` references together can be thought of as a _compile-time_
+/// version of a read-write lock.
+///
+///
+/// [`Sync`]: core::marker::Sync
+#[unstable(feature = "exclusive_wrapper", issue = "none")]
```
Comment (m-ou-se) [inline library/core/src/sync/exclusive.rs:75]: Can you [open a tracking issue](https://github.com/rust-lang/rust/issues/new?assignees=&labels=C-tracking-issue%2C+T-libs&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX)? Thanks!

Comment (guswynn) [reply library/core/src/sync/exclusive.rs:75]: yes!

**Comment by guswynn** [comment]
@m-ou-se https://github.com/rust-lang/rust/issues/98407 created and linked to the `unstable` attributes in this pr!
